File: 2022/day-05/python/part1_sol.py
f = open("../input.txt")
data = f.read()

starting_stacks, rearrangement_steps = data.split("\n\n")
# print(starting_stacks)
# print(rearrangement_steps)

stack_rows = starting_stacks.split("\n")

# find no of stacks
num_stacks = len(list(filter(lambda x: x.isdigit(), stack_rows[-1].split(" "))))

map = {}
for i in range(num_stacks):
    map[i+1] = []

for i in reversed(range(len(stack_rows)-1)):
    for j in range(1, len(stack_rows[i]), 4):
        map[j//4 + 1].append(stack_rows[i][j])

for key in map:
    while map[key][-1] == " ":
        map[key].pop()

# print(map)

rearrangement_steps = rearrangement_steps.rstrip().split("\n")

for step in rearrangement_steps:
    num_items, from_, to = [int(c) for c in step.split(" ") if c.isdigit()]
    map[to] += list(reversed(map[from_]))[:num_items]
    map[from_] = map[from_][:-num_items]

for key in map:
    print(map[key][-1], end="")

File: 2022/day-05/rust/part1_sol.rs
use std::fs;

fn main() {
    let data = fs::read_to_string("../input.txt").expect("Couldn't read from file");
    let (starting_stacks, mut rearrangement_steps) = data.split_once("\n\n").unwrap();
    // println!("{} {}", starting_stacks, rearrangement_steps);

    let stack_rows: Vec<&str> = starting_stacks.split("\n").collect();
    let stack_numbers_row = stack_rows.last().unwrap();

    let num_stacks = stack_numbers_row.split(" ").filter(|x| x.parse::<u32>().is_ok()).count();
    let num_stacks = num_stacks as u32;

    let mut stacks: Vec<String> = Vec::new();
    
    for _ in 0..num_stacks {
        stacks.push(String::new());
    }

    for i in (0..stack_rows.len()-1).rev() {
        for j in (1..stack_rows[i].len()).step_by(4) {
            stacks[j/4] += &stack_rows[i][j..j+1];
        }
    }

    stacks = stacks.iter().map(|x| x.trim_end().to_string()).collect();

    // println!("{:?}", stacks);    // stacks before rearrangement

    rearrangement_steps = rearrangement_steps.trim_end();

    for step in rearrangement_steps.split("\n") {
        let numbers = step.split(" ").filter(|x| x.parse::<u32>().is_ok());
        let numbers: Vec<u32> = numbers.map(|x| x.parse::<u32>().unwrap()).collect();
        let (num_items, from, to) = (numbers[0] as usize, numbers[1] as usize - 1, numbers[2] as usize - 1);

        // for _ in 0..num_items {
        //     let ch = stacks[from].pop().unwrap();
        //     stacks[to].push(ch);
        // }

        let l = stacks[from].len();
        let mut crates = stacks[from].split_off(l-num_items);
        crates = crates.chars().rev().collect();
        stacks[to] += &crates;    
    }

    for stack in &stacks {
        print!("{}", stack.chars().last().unwrap());
    }

}